File: src/types/data/Obra.ts
import * as z from "zod";

import { ObraFormSchema } from "@/schemas";
import Arquivo from "./Arquivo";
import Cliente from "./Cliente";
import {
  Sondagem_Percussao,
  Sondagem_Rotativa,
  Sondagem_Trado,
} from "./Sondagem";

export type Obra = {
  id: number;
  cod_obra: string;
  num_obra: number;
  ano: number;
  data_inicio: Date;
  data_fim: Date;
  tipo_logo: string | null;
  logradouro: string;
  lote: string | null;
  quadra: string | null;
  bairro: string;
  cidade: string;
  uf: string;
  cep: string | null;
  complemento_logo: string | null;
};

// TODO: Deprecate Type
export type TableObra = {
  id: number;
  cod_obra: string;
  ano: number;
  endereco: string;
  bairro: string;
  cidade: string;
  uf: string;
  cliente: string;
  proprietario: string | null;
};

export type FormObra = z.infer<typeof ObraFormSchema>;

export type InsertionObra = Omit<
  FormObra,
  | "cep"
  | "lote"
  | "quadra"
  | "complemento_logo"
  | "cliente"
  | "proprietario"
  | "sondagem_percussao"
  | "sondagem_rotativa"
  | "sondagem_trado"
> & {
  cep: string | null;
  lote: string | null;
  quadra: string | null;
  complemento_logo: string | null;
  clienteId: number;
  proprietarioId: number | null;
};

export type EntryObra = Omit<Obra, "clienteId" | "proprietarioId"> & {
  arquivos: Arquivo[] | null;
  cliente: {
    id: number;
    nome: string;
  };
  proprietario: {
    id: number;
    nome: string;
  } | null;
};

export type FullObra = Obra & {
  cliente: Cliente;
  proprietario: Cliente | null;
  arquivos: Arquivo[];
  sondagem_percussao: Sondagem_Percussao | null;
  sondagem_trado: Sondagem_Trado | null;
  sondagem_rotativa: Sondagem_Rotativa | null;
};

File: server/src/models/entities/obra.py
from datetime import date
from typing import Optional
from sqlalchemy import CHAR, Date, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column

from src.config.constants import CEP_LENGTH, MAX_BAIRRO_LENGTH, MAX_CIDADE_LENGTH, MAX_COD_OBRA_LENGTH, MAX_COMPLEMENTO_LOGO_LENGTH, MAX_LOGRADOURO_LENGTH, MAX_LOTE_LENGTH, MAX_QUADRA_LENGTH, MAX_TIPO_LOGO_LENGTH, UF_LENGTH
from src.models.entities.base import Base


class Obra(Base):
    __tablename__ = 'Obra'

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    cod_obra: Mapped[str] = mapped_column(String(MAX_COD_OBRA_LENGTH), index=True, unique=True)
    num_obra: Mapped[int] = mapped_column(index=True)
    ano: Mapped[int] = mapped_column(index=True)
    data_inicio: Mapped[date] = mapped_column(Date(), index=True)
    data_fim: Mapped[date] = mapped_column(Date(), index=True)
    tipo_logo: Mapped[Optional[str]] = mapped_column(String(MAX_TIPO_LOGO_LENGTH)) # Optional for compatibility reasons
    logradouro: Mapped[str] = mapped_column(String(MAX_LOGRADOURO_LENGTH), index=True)
    lote: Mapped[Optional[str]] = mapped_column(String(MAX_LOTE_LENGTH))
    quadra: Mapped[Optional[str]] = mapped_column(String(MAX_QUADRA_LENGTH))
    bairro: Mapped[str] = mapped_column(String(MAX_BAIRRO_LENGTH), index=True)
    cidade: Mapped[str] = mapped_column(String(MAX_CIDADE_LENGTH), index=True)
    uf: Mapped[str] = mapped_column(CHAR(UF_LENGTH), index=True)
    cep: Mapped[Optional[str]] = mapped_column(CHAR(CEP_LENGTH))
    complemento: Mapped[Optional[str]] = mapped_column(String(MAX_COMPLEMENTO_LOGO_LENGTH), index=True)
    cliente_id: Mapped[int] = mapped_column(ForeignKey("Cliente.id"))
    proprietario_id: Mapped[Optional[int]] = mapped_column(ForeignKey("Cliente.id"))
    
    def __repr__(self) -> str:
        return f"<Obra | id={self.id} | cod_obra={self.cod_obra} | ano={self.ano} | data_inicio={self.data_inicio} | data_fim={self.data_fim} | logradouro={self.tipo_logo + " " + self.logradouro}>"


